lib/src/mt_rtcp.c

```diff
@@ -0,0 +1,290 @@
+  //     mt_udp_hdr));
+
+  // uint16_t seq_id = ntohs(rtp->seq_number);
+
+  /* TODO: update nack list
+   * 1. remove received item from nack list (rx->stat_rtp_retransmit_succ++)
+   * 2. calculate missing seq_id(s)
+   * 3. make nack item
+   * 4. insert to nack list
+   **/
+
+  rx->stat_rtp_received++;
+
+  return 0;
+}
+
+/* where to trigger this nack sending? */
+int mt_rtcp_rx_send_nack_packet(struct mt_rtcp_rx* rx) {
+  struct mtl_main_impl* impl = rx->parent;
+  enum mtl_port port = rx->port;
+  struct rte_mbuf* pkt;
+  struct rte_ipv4_hdr* ipv4;
+
+  pkt = rte_pktmbuf_alloc(mt_get_tx_mempool(impl, port));
+  if (!pkt) {
+    err("%s(%d), pkt alloc fail\n", __func__, port);
+    return -ENOMEM;
+  }
+
+  struct mt_udp_hdr* hdr = rte_pktmbuf_mtod(pkt, struct mt_udp_hdr*);
+  ipv4 = &hdr->ipv4;
+
+  rte_memcpy(hdr, &rx->udp_hdr, sizeof(*hdr));
+  ipv4->packet_id = rx->ipv4_packet_id++;
+
+  mt_mbuf_init_ipv4(pkt);
+  pkt->data_len = sizeof(*hdr);
+  pkt->pkt_len = pkt->data_len;
+
+  /* TODO: build the nack packet
+   * 1. fill rtcp header
+   * 2. walk through the nack list
+   * 3. get the retransmit requests
+   * 4. fill the FCIs
+   * 5. update mbuf len
+   **/
+
+  uint16_t send = mt_dev_tx_sys_queue_burst(impl, port, &pkt, 1);
+  if (send != 1) {
+    err("%s, failed to send nack packet\n", __func__);
+    rte_pktmbuf_free(pkt);
+    return -EIO;
+  }
+
+  rx->stat_nack_sent++;
+
+  return 0;
+}
+
+static int rtcp_tx_stat(void* priv) {
+  struct mt_rtcp_tx* tx = priv;
+  enum mtl_port port = tx->port;
+
+  notice("%s(%d,%s), rtp sent %u nack recv %u rtp retansmit %u\n", __func__, port,
+         tx->name, tx->stat_rtp_sent, tx->stat_nack_received,
+         tx->stat_rtp_retransmit_succ);
+  tx->stat_rtp_sent = 0;
+  tx->stat_nack_received = 0;
+  tx->stat_rtp_retransmit_succ = 0;
+  if (tx->stat_rtp_retransmit_drop) {
+    warn("%s(%d,%s), rtp retansmit fail %u\n", __func__, port, tx->name,
+         tx->stat_rtp_retransmit_drop);
+    tx->stat_rtp_retransmit_drop = 0;
+  }
+
+  return 0;
+}
+
+static int rtcp_rx_stat(void* priv) {
+  struct mt_rtcp_rx* rx = priv;
+  enum mtl_port port = rx->port;
+
+  notice("%s(%d,%s), rtp recv %u nack sent %u rtp retransmit %u\n", __func__, port,
+         rx->name, rx->stat_rtp_received, rx->stat_nack_sent,
+         rx->stat_rtp_retransmit_succ);
+  rx->stat_rtp_received = 0;
+  rx->stat_nack_sent = 0;
+  rx->stat_rtp_retransmit_succ = 0;
+
+  return 0;
+}
+
+struct mt_rtcp_tx* mt_rtcp_tx_create(struct mtl_main_impl* impl,
+                                     struct mt_rtcp_tx_ops* ops) {
+  struct mt_rtcp_tx* tx =
+      mt_rte_zmalloc_socket(sizeof(struct mt_rtcp_tx), mt_socket_id(impl, ops->port));
+  if (!tx) {
+    err("%s, failed to allocate memory for mt_rtcp_tx\n", __func__);
+    return NULL;
+  }
+
+  struct rte_ring* ring =
+      rte_ring_create(ops->name, ops->buffer_size, mt_socket_id(impl, ops->port),
```

Comment:
Since retransmission(dequeue) is triggered on NACK received, the ring may have multiple consumers.

Reply:
We should keep both the normal tx and retransmission are running in the same scheduler(sch) that single producer and consumer is safe enough